shared/src/main/scala/mlscript/Typer.scala

```diff
@@ -1021,6 +1021,70 @@ class Typer(var dbg: Boolean, var verbose: Bool, var explainErrors: Bool, val ne
         val desug = If(IfThen(lhs, rhs), S(Var("false")))
         term.desugaredTerm = S(desug)
         typeTerm(desug)
+      case App(f: Term, a @ Tup(fields)) if (fields.exists(x => x._1.isDefined)) =>
+        val f_ty = typeTerm(f)
+        val fun_ty: SimpleType = f_ty.unwrapProxies match {
+          case tv: TypeVariable =>
+            def getLowerboundFuns(tv: TypeVariable): List[FunctionType] = {
+              val res: List[FunctionType] = tv.lowerBounds.map(x => 
+                x.unwrapProvs match {
+                  case PolymorphicType(_, AliasOf(fun_ty @ FunctionType(_, _))) =>
+                    List(fun_ty)
+                  case tvv: TypeVariable =>
+                    getLowerboundFuns(tvv) 
+                  case _ =>
+                    err("match error", f.toLoc)
+                    Nil
+                }).flatten
+              res
+            }
+            val funs = getLowerboundFuns(tv)
+            funs match {
+              case x :: Nil => 
+                funs.head
+              case Nil =>
+                err("Cannot extract any function", f.toLoc)
```

Comment:
This error is never exercised in the test. Besides, it should give some **context** as to what's happening! How do you expect *end users* to understand the source of the error with such message?